Add campaign summary export

JCantu248 opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This change adds the capability to export click summaries for the campaigns in an assessment. The output of these new features are a click summary in both a text file format and a JSON file. This data consists of the number of emails sent, the number of unique users who responded, the rate of user clicks, and the total number of clicks returned. 

UPDATE 8/30: Per request with @BenBreaksThings, changes were made to improve the campaign summaries. They are now no longer being called campaign click summaries, but simply campaign summaries. The fields reported have expanded to include email subject, from address, and the campaign start and end dates. Some significant refactor happened as a result and the changes will have to be reviewed. 


<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

The PCA team runs a script against Cobalt-Strike, called cs_phish_parse that creates this report from a series of csv input files.  This script is run manually by operators, and it was requested that it gets ported to Gophish and included as part of the export process. The PCA team requested both the JSON and plain text file output of the report. 

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->

<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

The Gophish-export command was run against multiple assessments, with campaigns that contain clicks and those that do not. The values output in the final click summary report reflected the correct number of emails sent, clicked on, and the correct number of unique users. 

A test case was also written to test the function that counts the number of unique users that clicked on a campaign email. 

## 📷 Screenshots (if appropriate) ##

<!-- Remove this section and header if not needed -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] _All_ future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests pass.

JCantu248 committed:
Output a click summary report to JSON, text file, and console. Helper functions added to write to console and file using the same string, and to find a unique target in a collection of clicks.

JCantu248 committed:
Merge branch 'develop' of github.com:cisagov/gophish-tools into add_campaign_summary_export

Merging main develop branch back into add_campaign_summary_export

JCantu248 committed:
Merge branch 'develop' of github.com:cisagov/gophish-tools into add_campaign_summary_export

Merging latest develop into feature branch. This should take the version number to 0.0.5

JCantu248 committed:
Bump project version from 0.0.5 to 0.0.6

JCantu248 committed:
Bump project version from 0.0.6 to 0.0.6-rc.1

JCantu248 committed:
Merge branch 'develop' of github.com:cisagov/gophish-tools into add_campaign_summary_export

Merging the bug fix for datetime selection not appearing into this branch. It will be included as release 0.0.6

JCantu248 committed:
Bump project version from 0.0.6-rc.1 to 0.0.6-rc.2

JCantu248 committed:
Caught a bug in export_user_reports when no clicks are recorded for a campaign. Testing correction and write_campaign_click_summary

JCantu248 committed:
Bump project version from 0.0.6-rc.2 to 0.0.6-rc.3

JCantu248 committed:
Bump project version from 0.0.6-rc.3 to 0.0.6-rc.4

JCantu248 committed:
Adding some fixes from testing involving dictionary field names.

JCantu248 committed:
Bump project version from 0.0.6-rc.5 to 0.0.6-rc.6

JCantu248 committed:
Bump project version from 0.0.6-rc.6 to 0.0.6-rc.7

JCantu248 committed:
Refactoring adding a list of campaign dicts to the campaign summary dict for JSON output

JCantu248 committed:
Bump project version from 0.0.6-rc.8 to 0.0.6-rc.9

JCantu248 committed:
Bump project version from 0.0.6-rc.9 to 0.0.6-rc.10
Small typo in output string was found and corrected here. Ready for PR.

JCantu248 committed:
Adding test case for finding the count of unique users who respond to a phishing campaign.

JCantu248 committed:
Update documentation for assessment click summary JSON objects, and fixed a field name in the source code

JCantu248 committed:
Bump project version from 0.0.6-rc.10 to 0.0.6

nickviola commented on src/tools/_version.py:
```diff
@@ -1,2 +1,2 @@
 """This file defines the version of this module."""
-__version__ = "0.0.6"
+__version__ = "0.0.6-rc.1"
```
Does this need to be updated or is the rc.1 tag valid here?

JCantu248 replied:
It should be finalized.

nickviola replied:
👍 referencing the commit:

resolved in 7ed1f98ec5c4396ffcdf131635af500f38627962

JCantu248 committed:
Bump tools version from 0.0.6-rc.1 to 0.0.6

nickviola reviewed (approved):
👍 LGTM. 

This is outside of the scope of this PR in my opinion but something we should consider for future cleanup/improvements is to circle back and create an issue to clean up all outputs in this project. For example, where we are using logging along with print to dump to stdout and try to make it consistent. Same for any file output paths we can standardize.

jsf9k commented:
>  LGTM.
> 
> This is outside of the scope of this PR in my opinion but something we should consider for future cleanup/improvements is to circle back and create an issue to clean up all outputs in this project. For example, where we are using logging along with print to dump to stdout and try to make it consistent. Same for any file output paths we can standardize.

Great point @nickviola!  I suggest using logging everywhere.  Please make an issue for this so the idea is captured and not forgotten.

nickviola commented:
> > LGTM.
> > This is outside of the scope of this PR in my opinion but something we should consider for future cleanup/improvements is to circle back and create an issue to clean up all outputs in this project. For example, where we are using logging along with print to dump to stdout and try to make it consistent. Same for any file output paths we can standardize.
> 
> Great point @nickviola! I suggest using logging everywhere. Please make an issue for this so the idea is captured and not forgotten.

👍 I agree and also prefer using logging exclusively. I will get an issue created to keep it on our radar. Thanks for the feedback @jsf9k!

dav3r commented:
> > > LGTM.
> > > This is outside of the scope of this PR in my opinion but something we should consider for future cleanup/improvements is to circle back and create an issue to clean up all outputs in this project. For example, where we are using logging along with print to dump to stdout and try to make it consistent. Same for any file output paths we can standardize.
> > 
> > 
> > Great point @nickviola! I suggest using logging everywhere. Please make an issue for this so the idea is captured and not forgotten.
> 
> 👍 I agree and also prefer using logging exclusively. I will get an issue created to keep it on our radar. Thanks for the feedback @jsf9k!

To close the loop here, I just wanted to mention that @nickviola created #41 for this. 👍

dav3r commented:
@JCantu248 - Can you please clean up this incomplete sentence in the "Testing" section above?  I'm dying to know about this test case! 😃 
> A test case was also written to test the c

dav3r commented on README.md:
```diff
@@ -183,6 +183,20 @@ campaign.
 | first_report | First report (click) generated by a targeted user. Format: "YYYY-MM-DDThh:mm.ss" | datetime |
 | total_num_reports | Total number of user reports received during a campaign. | integer |
 
+## Assessment Click Summary ##
```
Update verbiage to be consistent with `User Report Field Dictionary` section above:
```suggestion
## Assessment Click Summary Field Dictionary ##
```

JCantu248 replied:
See 5b4969a6f0a931b2c0d60a69662b1db6bf6a31c0 for changes

dav3r replied:
All of the other sections say "Field Dictionary" - you put "Field Description".

JCantu248 replied:
Oops...check 06e366bb0effe1ce8ff6ecae40834ea46dc296a1

dav3r commented on README.md:
```diff
@@ -183,6 +183,20 @@ campaign.
+| total_emails_sent| The total number of emails sent by this campaign | integer |
+| unique_user_count | The number of unique users that clicked on an email sent by the campaign| integer |
+| click_rate | The ratio of emails clicked and total emails sent | float |
+| total_clicks | The total number of clicks reported by a campaign | integer |
```
End all descriptions with spaces, to be consistent with other sections:
```suggestion
| num_campaigns | Number of campaigns in this assessment. | integer |
| campaign_id | The ID number of the campaign. | string |
| total_emails_sent| The total number of emails sent by this campaign. | integer |
| unique_user_count | The number of unique users that clicked on an email sent by the campaign. | integer |
| click_rate | The ratio of emails clicked and total emails sent. | float |
| total_clicks | The total number of clicks reported by a campaign. | integer |
```

JCantu248 replied:
See 5b4969a6f0a931b2c0d60a69662b1db6bf6a31c0 for changes

dav3r commented on README.md:
```diff
@@ -183,6 +183,20 @@ campaign.
+| campaign_id | The ID number of the campaign | string |
+| total_emails_sent| The total number of emails sent by this campaign | integer |
+| unique_user_count | The number of unique users that clicked on an email sent by the campaign| integer |
+| click_rate | The ratio of emails clicked and total emails sent | float |
```
The `click_rate` isn't really a ratio.  It's the percentage of unique users that clicked on an email out of the total number of emails sent.

JCantu248 replied:
See 5b4969a6f0a931b2c0d60a69662b1db6bf6a31c0 for changes

dav3r commented on src/models/models.py:
```diff
@@ -250,3 +250,32 @@ def parse(cls, json):
+
+
+class Click(Model):
+    """The Click class."""
```
This is not a particularly useful docstring- can you please try to improve it a bit?  For some inspiration, you can look [here](https://www.programiz.com/python-programming/docstrings) (see `Example 6: Docstrings for Python class`).

JCantu248 replied:
See 10ea3c85dde6efaf4d510f4c250b74a06fc92685

dav3r replied:
Much better, thanks!

jsf9k commented on src/util/double_print.py:
```diff
@@ -0,0 +1,7 @@
+"""Utility function to print out to stdout and file at the same time."""
+
+
+def double_print(outfile, line_string):
+    """Write to console and a file using the same string."""
+    print(line_string)
+    outfile.write(line_string + "\n")
```
Prefer Python standard library logging to this kind of thing.  See #41.

JCantu248 replied:
See 0a3f4bc0587431f616a0d35548f9fc80da72a16b for changes to the way we are outputting. I implemented a logger with a handler for file output as well as stdout.

jsf9k replied:
Sorry I lost the thread here for a while, but thanks for taking care of that!

dav3r commented on src/tools/gophish_export.py:
```diff
@@ -253,6 +254,74 @@ def get_application(rawEvent):
+                campaign_click_stats["unique_user_count"]
+            ) / float(campaign_click_stats["total_emails_sent"])
+        else:
+            campaign_click_stats["click_rate"] = 0
```
Ensure `campaign_click_stats["click_rate"]` is always a `float`.
```suggestion
            campaign_click_stats["click_rate"] = 0.0
```

JCantu248 replied:
See 0a3f4bc0587431f616a0d35548f9fc80da72a16b for changes to this line

dav3r replied:
Looks good, though in the future, it's best practice to put a fix like this in its own commit, rather than hide it within a larger commit that addresses other issues.  Better yet, in this case, it would have been simplest to just use the "Commit suggestion" feature in the GitHub web GUI.

dav3r commented on src/tools/gophish_export.py:
```diff
@@ -253,6 +254,74 @@ def get_application(rawEvent):
+    campaigns = list()
+    campaign_ids = get_campaign_ids(api, assessment_id)
+    num_campaigns = len(campaign_ids)
+    summary_json = assessment_id + "_click_summary.json"
+    summary_outfile_name = assessment_id + "_click_summary.txt"
+    summary_outfile = open(summary_outfile_name, "w")
```
Perhaps something like this to improve the readability of the variable names?  Obviously you'd also need to update these variable names throughout this function if you choose to accept this suggestion.

```suggestion
    json_filename = assessment_id + "_click_summary.json"
    text_filename = assessment_id + "_click_summary.txt"
    summary_outfile_text = open(text_filename, "w")
```

dav3r replied:
I still would like to see better variable names here.  As of https://github.com/cisagov/gophish-tools/pull/40/commits/0a3f4bc0587431f616a0d35548f9fc80da72a16b, you have them called `json_filename` and `summary_outfile_name`.  I'd rather see more similarly-named variables (e.g. `json_filename` / `text_filename` or `summary_json_filename` / `summary_text_filename`).

dav3r replied:
These variable names were cleaned up in https://github.com/cisagov/gophish-tools/pull/40/commits/0a0c1649c0f6a2b3debf21bc8f1682c4eb37a8d5 - thanks!

dav3r commented on tests/conftest.py:
```diff
@@ -13,7 +13,16 @@
 # cisagov Libraries
-from models.models import SMTP, Assessment, Campaign, Group, Page, Target, Template
+from models.models import (
+    SMTP,
+    Assessment,
+    Campaign,
+    Click,
+    Group,
+    Page,
+    Target,
+    Template,
```
We always prefer to alphabetize wherever possible:

```suggestion
    Assessment,
    Campaign,
    Click,
    Group,
    Page,
    SMTP,
    Target,
    Template,
```

dav3r replied:
@JCantu248 - Reminder that this item still has not been addressed.

dav3r replied:
@JCantu248 - Pinging again on this small alphabetization issue.

JCantu248 replied:
It looks like isort in pre-commit will complain with the requested changes to the import list and revert them back to the original way.

dav3r replied:
That is strange, but if that is what `isort` is doing, we can resolve this for now.

@mcdonnnj this oddness is something that you might be interested in looking into at some point in the future.

dav3r commented on tests/conftest.py:
```diff
@@ -298,6 +307,50 @@ def email_target_json():
+            Click(
+                message="Testing",
+                user="jane.smith1@domain.tld",
+                source_ip="8.8.8.8",
```
Avoid using public IP addresses in test code when possible:
```suggestion
                source_ip="10.0.0.0",
```

JCantu248 replied:
Changed in f5ee8f21d3d84343e41b0833e5140462593b9424

dav3r commented on tests/conftest.py:
```diff
@@ -298,6 +307,50 @@ def email_target_json():
+            Click(
+                message="Testing",
+                user="john.doe1@domain.tld",
+                source_ip="8.8.8.8",
```
Avoid using public IP addresses in test code when possible:
```suggestion
                source_ip="10.0.0.1",
```

JCantu248 replied:
Changed in f5ee8f21d3d84343e41b0833e5140462593b9424

dav3r commented on tests/conftest.py:
```diff
@@ -298,6 +307,50 @@ def email_target_json():
+            Click(
+                message="Testing",
+                user="jane.smith2@domain.tld",
+                source_ip="8.8.8.8",
```
Avoid using public IP addresses in test code when possible:
```suggestion
                source_ip="10.0.0.2",
```

JCantu248 replied:
Changed in f5ee8f21d3d84343e41b0833e5140462593b9424

dav3r commented on tests/conftest.py:
```diff
@@ -298,6 +307,50 @@ def email_target_json():
+            Click(
+                message="Testing",
+                user="john.doe2@domain.tld",
+                source_ip="8.8.8.8",
```
Avoid using public IP addresses in test code when possible:
```suggestion
                source_ip="10.0.0.3",
```

JCantu248 replied:
Changed in f5ee8f2

dav3r commented on tests/test_tools.py:
```diff
@@ -43,9 +47,15 @@ def test_assessment_exists_found(self, mock_api, multiple_campaign_object):
+    @patch("tools.connect")
+    def test_find_unique_target_clicks_count(self, mock_api, multiple_click_object):
+        """Verify that the correct number of unique users in a click list is found."""
+        clicks = multiple_click_object
+
+        assert find_unique_target_clicks_count(clicks) == 4
```
This can be simplified:
```suggestion
        assert find_unique_target_clicks_count(multiple_click_object) == 4
```

JCantu248 replied:
Changed in 1159e343852e32e59f95b4a6a4a86bf8dd3d454b

dav3r reviewed (changes requested):
Please take a look at my feedback and suggestions - let me know if you have any questions.

JCantu248 committed:
Changes to gophish-export. Removing double print and adding file handler to logger. Changes to variable names and other minor fixes.

JCantu248 committed:
Changes to README.md. Punctuation fixds and other small changes.

JCantu248 committed:
Change the Click model class to have a more descriptive docstring

JCantu248 committed:
Changes to conftest. Put models imports into alphabetical order, and changed test IP addresses from public addresses to private.

JCantu248 committed:
Simplified code for test_find_unique_target_clicks.

JCantu248 committed:
Bump project version from 0.0.6 to 0.0.6-rc.1

JCantu248 commented:
> @JCantu248 - Can you please clean up this incomplete sentence in the "Testing" section above? I'm dying to know about this test case! 😃
> 
> > A test case was also written to test the c

Fixed!

JCantu248 committed:
Changed User Report Field Description to field Dictionary

JCantu248 committed:
Setting version to RC.11 for testing purposes. Was previously finalized prior to creating the PR, but its going back in for testing. Manually setting the version to rc.11

JCantu248 committed:
Bump project version from 0.0.6-rc.11 to 0.0.6-rc.12

JCantu248 committed:
Bug fix to force overwrite of existing summary file

JCantu248 committed:
Bump project version from 0.0.6-rc.12 to 0.0.6-rc.13

JCantu248 committed:
Manually deleting previous summary.txt file if already exists

JCantu248 committed:
Bump project version from 0.0.6-rc.13 to 0.0.6-rc.14

dav3r commented on src/tools/gophish_export.py:
```diff
@@ -253,6 +254,71 @@ def get_application(rawEvent):
+    num_campaigns = len(campaign_ids)
+    json_filename = assessment_id + "_click_summary.json"
+    summary_outfile_name = assessment_id + "_click_summary.txt"
+    if os.path.exists(summary_outfile_name):
+        os.remove(summary_outfile_name)
+    fh = logging.FileHandler(summary_outfile_name, "w+")
```
If `summary_outfile_name` exists and you remove it, then there should no longer be any need for "w+" mode.
```suggestion
    if os.path.exists(summary_outfile_name):
        os.remove(summary_outfile_name)
    fh = logging.FileHandler(summary_outfile_name, "w")
```

dav3r replied:
Looks like this was corrected back in https://github.com/cisagov/gophish-tools/commit/0a0c1649c0f6a2b3debf21bc8f1682c4eb37a8d5#diff-56cd9a15b1d8135c0ca3f40187af79b1c273eed9a40dbc3c8f6d43705925674bL275-L276 - thanks!

JCantu248 committed:
Major changes. Revised the script to a new JSON format discussed by the PCA Dev team.

JCantu248 committed:
Bump project version from 0.0.6-rc.14 to 0.0.6-rc.15

JCantu248 committed:
Bug fix for concatenation with string and date objects when creating the summary file.

Using strftime() to parse a datetime object and convert to a string for concatenation purposes

JCantu248 committed:
Bump project version from 0.0.6-rc.15 to 0.0.6-rc.16

JCantu248 committed:
Continuing the bug fixes

JCantu248 committed:
Refactor function to not use template for building JSON dict.
JSON is being built out entirely in the function. This saves us some potential permission issues later by not needing a template file for the JSON structure

JCantu248 committed:
Fixing API bug calling API to get campaign

JCantu248 committed:
Bump project version from 0.0.6-rc.17 to 0.0.6-rc.18

JCantu248 committed:
Bump project version from 0.0.6-rc.18 to 0.0.6-rc.19

JCantu248 committed:
Revert "Continuing the bug fixes"

This reverts commit 644c76f25bfb27338b5b908e65e6888a78f60cf0.

JCantu248 committed:
Bump project version from 0.0.6-rc.19 to 0.0.6-rc.20

JCantu248 committed:
Still debugging problems in opening campaign template file

JCantu248 committed:
Bump project version from 0.0.6-rc.21 to 0.0.6-rc.22

JCantu248 committed:
Working through API issues

JCantu248 committed:
Bump project version from 0.0.6-rc.23 to 0.0.6-rc.24

JCantu248 committed:
Debugging API issues

JCantu248 committed:
Still debugging. Switching back to using campaign data template by request

JCantu248 committed:
Fixing bug related to receiving a str when trying to parse a datetime.

Datetime parse is not needed and removed.

JCantu248 committed:
Changing logger file out to just a normal file out.

The logger was prepending all messages to the file with INFO. This was undesirabl for a report, so it was changed back to doing just a regular file out. The contents of the campaign summaries are no longer output to the console, only to file.

JCantu248 committed:
Adding newline characters in output messages for readability

JCantu248 committed:
Final last minute fixes and finalize work

JCantu248 committed:
Bump project version from 0.0.6-rc.29 to 0.0.6

JCantu248 committed:
Fix newline character prior to row of hyphens.

Trying to improve readability and separate the campaigns in an assessment

JCantu248 committed:
Updating README.md to account for changes to JSON.

Campaign click summary has been renamed to Campaign Summary, along with changes to the reported fields. The README.md file has been updated to account for these changes.

JCantu248 committed:
Removing unnecessary campaign variable declaration.

The variable was declared at the start of the function, and then initialized at line 274. The first declaration is unnecessary and removed.

dav3r commented on README.md:
```diff
@@ -183,6 +183,22 @@ campaign.
+
+| Name | Description | Type |
+|------|-------------|:----:|
+| subject | The Subject line for a Gophish generated email. | string |
```
Remove unnecessary capitalization:
```suggestion
| subject | The subject line for a Gophish generated email. | string |
```

jsf9k replied:
I'd make one other change here as well, if you're going to be touching it anyway:
```suggestion
| subject | The subject line for a Gophish-generated email. | string |
```

JCantu248 replied:
See cdd706a3b4b8734351f7904e8fb4a43dd83fc8ba for suggested changes applied.

dav3r commented on src/tools/gophish_export.py:
```diff
@@ -253,6 +253,86 @@ def get_application(rawEvent):
+    for click in clicks:
+        if click["user"] not in uniq_list:
+            uniq_list.append(click["user"])
+    return len(uniq_list)
```
I would just use a set here instead of list, then you don't need the `if` statement:
```suggestion
    unique_users = set()
    for click in clicks:
        unique_users.add(click["user"])
    return len(unique_users)
```

JCantu248 replied:
See 5816a24f37e9a600f2b55946de18b7acb8d0da07 for revisions

dav3r commented on src/tools/gophish_export.py:
```diff
@@ -253,6 +253,86 @@ def get_application(rawEvent):
+    """Output a campaign summary report to JSON, console, and a text file."""
+    campaign_ids = get_campaign_ids(api, assessment_id)
+    campaign_data_template = "campaign_data.json"
+    campaign_summary_json = assessment_id + "_campaign_data.json"
+    campaign_summary_textfile = (
+        assessment_id + "_summary_" + str(datetime.now()) + ".txt"
+    )
```
f-strings are preferable just about everywhere (except in logging statements).  
Also, do you really want that space (from `str(datetime.now())` in your filenames?  I'd use the same format as you use later on:
```suggestion
    campaign_summary_json = f"{assessment_id}_campaign_data.json"
    campaign_summary_textfile = f"{assessment_id}_summary_{datetime.strftime(datetime.now(), '%Y-%m-%dT%H:%M:%S')}.txt"
```

JCantu248 replied:
The reason behind the use of datetime was by the desire to keep a separate file for each time the export is run. They did not want repeated runs of the export script to result in the outputted text file being overwritten.

dav3r replied:
> The reason behind the use of datetime was by the desire to keep a separate file for each time the export is run. They did not want repeated runs of the export script to result in the outputted text file being overwritten.

I don't have any issue with putting a timestamp in the filename- I just try to avoid creating filenames with spaces in them.

JCantu248 replied:
See 5816a24f37e9a600f2b55946de18b7acb8d0da07

dav3r commented on src/tools/gophish_export.py:
```diff
@@ -253,6 +253,86 @@ def get_application(rawEvent):
+        elif campaign.name.endswith("_level-6"):
+            level = "level-6"
+        else:
+            continue
```
Do you want to log a warning message here if a campaign name doesn't end with `_level-n`?

JCantu248 replied:
See 5816a24f37e9a600f2b55946de18b7acb8d0da07

dav3r replied:
```
  logging.warn(
      "Encountered campaign (%s) that is unable to be processed for campaign summary export. "
      "Skipping campaign",
      campaign.name,
  )
```
This is good, but it probably should also mention why the campaign is unable to be processed.

dav3r replied:
I see the updated error message in https://github.com/cisagov/gophish-tools/pull/40/commits/7475539f71d5638f3d383d3fa13120d221ca274d - thanks!

JCantu248 replied:
See 5816a24f37e9a600f2b55946de18b7acb8d0da07 for changes to this file

dav3r commented on src/tools/gophish_export.py:
```diff
@@ -253,6 +253,86 @@ def get_application(rawEvent):
+        if total_clicks > 0:
+            percent_clicks = unique_clicks / float(total_clicks)
+        else:
+            percent_clicks = 0
```
Ensure `percent_clicks` is always a `float`:
```suggestion
            percent_clicks = 0.0
```

JCantu248 replied:
See 5816a24f37e9a600f2b55946de18b7acb8d0da07

dav3r commented on src/tools/gophish_export.py:
```diff
@@ -253,6 +253,86 @@ def get_application(rawEvent):
+    with open(campaign_data_template) as template:
+        campaign_data = json.load(template)
+
+    logging.info("Writing campaign summary report to %s" % campaign_summary_textfile)
```
Use preferred formatting syntax in logging functions:
```suggestion
    logging.info("Writing campaign summary report to %s", campaign_summary_textfile)
```

JCantu248 replied:
See 5816a24f37e9a600f2b55946de18b7acb8d0da07

dav3r commented on src/tools/gophish_export.py:
```diff
@@ -253,6 +253,86 @@ def get_application(rawEvent):
+        )
+
+    file_out.close()
+    logging.info("Writing out summary JSON to %s" % campaign_summary_json)
```
Use preferred formatting syntax in logging functions:
```suggestion
    logging.info("Writing out summary JSON to %s", campaign_summary_json)
```

JCantu248 replied:
See 5816a24f37e9a600f2b55946de18b7acb8d0da07

dav3r commented on tests/conftest.py:
```diff
@@ -298,6 +307,50 @@ def email_target_json():
+def multiple_click_object():
+    """Return a list of clicks to match the correct number of unique users."""
+    clicks = list()
+    for x in range(1, 2):
```
What is the purpose of this `for x in range(1, 2):` statement here?

JCantu248 replied:
Edit: I added that loop in there to produce a duplicate of the data I was testing. I wanted two copies of each click object so that find_unique_target_clicks_count() would be able to discern from duplicate objects in the set. This way I would be able to provide two of every object (8 click objects total) but I should be expecting a result of 4 after duplicates are disregarded.

dav3r replied:
Got it, except that loop only has a single iteration- it sounds like you want `for x in range(0, 2):`.

dav3r replied:
Looks like you corrected this in https://github.com/cisagov/gophish-tools/pull/40/commits/3bceac856e75b598686b2aef04c4ce05f86790d8 - thanks!

JCantu248 replied:
See 3bceac856e75b598686b2aef04c4ce05f86790d8 for changes to this file.

dav3r reviewed (changes requested):
Please take a look at my feedback - thanks!

Also, reminder that [this conversation](https://github.com/cisagov/gophish-tools/pull/40/files#r682104597) is still unresolved.

JCantu248 committed:
Updating README.md. Removing capitalized letters and hyphentating Gophish-generated, in response to PR feedback.

JCantu248 committed:
Revisions to gophish_export.py

Changes to logging formats, new warning log for unprocessable campaigns, and changing find_unique_user_count() to use a set over a list.

jsf9k commented on src/tools/gophish_export.py:
```diff
@@ -253,6 +253,88 @@ def get_application(rawEvent):
+    file_out = open(campaign_summary_textfile, "w+")
+    file_out.write("Campaigns for Assessment: " + assessment_id)
+
+    for campaign_id in campaign_ids:
+        campaign = api.campaigns.get(campaign_id)
+        if campaign.name.endswith("_level-1"):
+            level = "level-1"
+        elif campaign.name.endswith("_level-2"):
+            level = "level-2"
+        elif campaign.name.endswith("_level-3"):
+            level = "level-3"
+        elif campaign.name.endswith("_level-4"):
+            level = "level-4"
+        elif campaign.name.endswith("_level-5"):
+            level = "level-5"
+        elif campaign.name.endswith("_level-6"):
+            level = "level-6"
+        else:
```
Since you're touching this code anyway, this logic could me made less repetitive and less error-prone if other levels are ever added by using a regex.  I'm thinking of something like this:
 ```suggestion
    regex = re.compile(r"^.*_(?P<level>level-[1-6])$")
    for campaign_id in campaign_ids:
        campaign = api.campaigns.get(campaign_id)
        match = regex.fullmatch(campaign.name)
        if match:
          level = match.group("level")
        else:
```

JCantu248 replied:
See 5816a24f37e9a600f2b55946de18b7acb8d0da07 for changes to this file

jsf9k reviewed (commented):
I made one suggestion for improvement.

JCantu248 committed:
Bug fix in conftest.

Should have been iterating twice over adding click objects to a test data set, but accidentally let it iterate once.

JCantu248 committed:
Changing export campaign summary level matching logic.

Changing error prompt to be more useful, and switching if-else block to regex

JCantu248 committed:
Merging latest commits in main develop branch into development branch prior to mergeing add_campaign_summary_export

JCantu248 committed:
Pre-release candidate for testing after recent changes.

JCantu248 committed:
Bump project version from 0.0.6-rc.30 to 0.0.6

JCantu248 committed:
Update tests/conftest.py to alphabetize imports.

Doing a commit suggestion here in Github. Looks like pre-commit hooks on my local branch will overwrite the requested changes and revert back to the incorrect way.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

JCantu248 committed:
Revert "Bump project version from 0.0.6-rc.30 to 0.0.6"

This reverts commit a1667916bcb00d9e0442f90441ab91f1e71a4d2d. Reverting changes to import. Leaving conftests imports back to the way isort wants them to be

dav3r reviewed (approved):
Thanks for replying to all of my feedback and questions!  🏎️
